### PR TITLE
SPOI-11409 Added HDFS parallel file copy application template

### PR DESCRIPTION
**Pull request by chaithu14**
@shubham-pathak22 Please review and merge.

### CODE REVIEW

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.apex.malhar.lib.io.block.PartFileWriter;
+import org.apache.hadoop.conf.Configuration;
+
+@ApplicationAnnotation(name="Hdfs-to-Hdfs-parallel-file-copy-App")
+public class Application implements StreamingApplication
+{
+  @Override
+  public void populateDAG(DAG dag, Configuration conf)
+  {
+    // Add FSInputModule as input and PartFileWriter as output operators respectively to dag.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: "respectively" not needed

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.StreamingApplication;
+import com.datatorrent.api.annotation.ApplicationAnnotation;
+import com.datatorrent.lib.io.fs.FSInputModule;
+
+import org.apache.apex.malhar.lib.io.block.PartFileWriter;
+import org.apache.hadoop.conf.Configuration;
+
+@ApplicationAnnotation(name="Hdfs-to-Hdfs-parallel-file-copy-App")
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: "parallel file copy" suggests we are copying multiple files parallelly. But what we are essentially doing is copying part files. I suggest the name must be intuitive.

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+public class Application implements StreamingApplication
+{
+  @Override
+  public void populateDAG(DAG dag, Configuration conf)
+  {
+    // Add FSInputModule as input and PartFileWriter as output operators respectively to dag.
+    FSInputModule input = dag.addModule("HDFSInputModule", new FSInputModule());
+    PartFileWriter output = dag.addOperator("PartFileCopy", new PartFileWriter());
+
+    // Create a stream for blocks, metadata from Input to PartFileWriter
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: blocks, metadata  ? 
Do you want to say block metadata OR blockMetadata, blockData, fileMetaData ?

You could simply write
Add streams for blockMetadata, blockData, fileMetaData

File: hdfs-parallel-file-copy/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+### Description
+The HDFS to HDFS parallel file copy Application Template continuously ingests files as blocks from one Hadoop cluster to another for backup or further processing without worrying about the compaction.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: The name "parallel file copy" is not intuitive

File: hdfs-parallel-file-copy/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- The application keeps the same data at destination but create part files instead of complete file.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: create -> creates

File: hdfs-parallel-file-copy/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+### Description
+The HDFS to HDFS parallel file copy Application Template continuously ingests files as blocks from one Hadoop cluster to another for backup or further processing without worrying about the compaction.
+- The application keeps the same data at destination but create part files instead of complete file. 
+- The application scales linearly with number of block readers.
+- The application is fault tolerant and can withstand node and cluster outages without data loss.
+- The application is also highly performance and can perform as fast as the network bandwidth allows.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: performance -> performant

File: hdfs-parallel-file-copy/README.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+- The only configuration user needs to provide is source and destination HDFS paths.
+- This enterprise grade application template will dramatically reduce your time to market and cost of operations.
+
+Import the application from DataTorrent AppHub and launch it to ingest data from one Hadoop cluster to another. Follow the tutorial videos or walkthrough document below to launch the template and run it.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: Not necessary "from one Hadoop cluster to another" . You may skip this
Just mention, Import the app from.. and follow the tutorials...

Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: Would be better if we add a link to *DataTorrent AppHub * to point it to the actual page

File: hdfs-parallel-file-copy/README.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+### Jumpstart
+Import and run application template as an operable proof of concept. Please watch the [walkthrough video](Update link location here) to import and launch the application.
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: the application

File: hdfs-parallel-file-copy/README.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+<iframe src="Update link location here" allowfullscreen="allowfullscreen" class="video" id="basicVideo" ga-track="basicVideo"></iframe>
+
+### Logical Plan
+
+Here is a preview of the logical plan of the application template
+
+![Logical Plan](Update Link here)
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: Add screenshots

File: hdfs-parallel-file-copy/README.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+### Launch App Properties
+
+Here is a preview of the properties to be set at application launch
+
+![Launch App Properties](Update Link here)
```
Comment (shubham-pathak22) [inline hdfs-parallel-file-copy/README.md]: Add screen shots

File: hdfs-parallel-file-copy/README.md (hunk 4 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Done

File: hdfs-parallel-file-copy/README.md (hunk 5 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Done.

File: hdfs-parallel-file-copy/README.md (hunk 6 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Done

File: hdfs-parallel-file-copy/README.md (hunk 7 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Done.

Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: We can only download from *DataTorrent AppHub*, but in the description it was import. So, i didn't added the link. I didn't see the link for the other app templates. I would prefer to make it uniform.

File: hdfs-parallel-file-copy/README.md (hunk 8 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Done.

File: hdfs-parallel-file-copy/README.md (hunk 9 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Sceenshots will be added after the walkthrough document and video recording.

File: hdfs-parallel-file-copy/README.md (hunk 10 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/README.md]: Same here.

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java (hunk 2 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: Done.

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java (hunk 1 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: Done.

File: hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java (hunk 3 above)
Comment (chaithu14) [inline hdfs-parallel-file-copy/src/main/java/com/datatorrent/apps/Application.java]: Done.

**Comment by chaithu14** [comment]
Removed the PartFileWriter from app template and updated the PR.

**Pull request merged**